odl/trafos/wavelet.py

```diff
@@ -21,9 +21,836 @@
+    from a given wavelet coefficient list.
+    Utilizes a PyWavelet function ``pywt.dwtn``
+
+    Parameters
+    ----------
+        coeff_list: : `list`
+            A list of wavelet approximation and detail coefficients
+            organized in the following way
+            [caaaN, (aadN, adaN, addN, daaN, dadN, ddaN, dddN), ...
+            (aad1, ada1, add1, daa1, dad1, dda1, ddd1)].
+            The appreviations refer to,
+
+            aaa = approx. on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+            aad = approx. on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+            ada = approx. on 1st dim, detail on 3nd dim, approx. on 3rd dim,
+
+            add = approx. on 1st dim, detail on 3nd dim, detail on 3rd dim,
+
+            daa = detail on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+            dad = detail on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+            dda = detail on 1st dim, detail on 2nd dim, approx. on 3rd dim,
+
+            ddd = detail on 1st dim, detail on 2nd dim, detail on 3rd dim,
+
+            N = the number of scaling levels
+        wbasis :  ``_pywt.Wavelet``
+            Describes properties of a selected wavelet basis
+        mode : `str`
+            Signal extention mode. For possible extensions see
+            ``pywt.MODES.modes``
+        nscales : `int`
+            Number of scales in the coefficient list
+
+    Returns
+    -------
+        x : `numpy.ndarray`.
+        A wavalet reconstruction.
+
+    See also
+    --------
+    pywt : http://www.pybytes.com/pywavelets/
+    """
+    aaa = coeff_list[0]
+    k = 1
+    (aad, ada, add, daa, dad, dda, ddd) = coeff_list[k]
+    coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
+                  'daa': daa, 'dad': dad, 'dda': dda, 'ddd': ddd}
+    for k in range(2, nscales+1):
+        aaa = pywt.idwtn(coeff_dict, wbasis, mode)
+        (aad, ada, add, daa, dad, dda, ddd) = coeff_list[k]
+        coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
+                      'daa': daa, 'dad': dad, 'dda': dda, 'ddd': ddd}
+
+    x = pywt.idwtn(coeff_dict, wbasis, mode)
+    return x
+
+
+class DiscreteWaveletTrafo(Operator):
+
+    """Discrete wavelet trafo between discrete L2 spaces."""
+
+    def __init__(self, dom, nscales, wbasis, mode):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        dom : :class:`~odl.discr.lp_discr.DiscreteLp`
+            Domain of the wavelet transform (the "image domain").
+            The exponent :math:`p` of the discrete :math:`L^p`
+            space must be equal to 2.0.
+        nscales : `int`
+            Number of scales in the coefficient list
+        wbasis :  ``_pywt.Wavelet``
+            Describes properties of a selected wavelet basis
+        mode : `str`
+            Signal extention mode. For possible extensions see
+            ``pywt.MODES.modes``
```

Comment:
I agree. Even more useful would be a direct hyperlink to the corresponding Pywavelets doc :-)